<a href="https://colab.research.google.com/github/punramon/python_dataviz2024/blob/main/Data_Viz_2024_Exam_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

ให้สืบหาฆาตรกร ในคดี ฆาตรกรรม (murder) ที่เกิดขึ้นในวันที่ 15 มกราคม 2018 ที่เมือง SQL City โดยเริ่มจากการตรวจสอบ ข้อมูลคดี (crime_scene_report.csv)

หมายเหตุ
* person.csv ข้อมูลประชากร
* interview.csv ข้อมูลการสอบสวน
* income.csv ข้อมูลรายได้
* get_fit_now_member.csv ข้อมูลสมาชิก ฟิตเนส Get Fit Now
* get_fit_now_check_in.csv ข้อมูลการเข้าใช้ ฟิตเนส Get Fit Now
* facebook_event_checkin.csv ข้อมูลการเช็คอิน Facebook
* driver_license.csv ข้อมูลใบขับขี่
* crime_scene_report.csv ข้อมูลคดี

## import data

In [ ]:
import pandas as pd

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
import os

path = '/content/drive/MyDrive/Exam_2' #set path --> folder and then join path เพื่อให้โค้ดrunได้ทุกที่

person_file_path = os.path.join(path,'person.csv')
interview_file_path = os.path.join(path,'interview.csv')
income_file_path = os.path.join(path,'income.csv')
get_fit_now_member_file_path = os.path.join(path,'get_fit_now_member.csv')
get_fit_now_check_in_file_path = os.path.join(path,'get_fit_now_check_in.csv')
facebook_event_checkin_file_path = os.path.join(path,'facebook_event_checkin.csv')
driver_license_file_path = os.path.join(path,'drivers_license.csv')
crime_scene_report_file_path = os.path.join(path,'crime_scene_report.csv')
print(person_file_path)
print(interview_file_path)
print(income_file_path)
print(get_fit_now_member_file_path)
print(get_fit_now_check_in_file_path)
print(facebook_event_checkin_file_path)
print(driver_license_file_path)
print(crime_scene_report_file_path)

/content/drive/MyDrive/Exam_2/person.csv
/content/drive/MyDrive/Exam_2/interview.csv
/content/drive/MyDrive/Exam_2/income.csv
/content/drive/MyDrive/Exam_2/get_fit_now_member.csv
/content/drive/MyDrive/Exam_2/get_fit_now_check_in.csv
/content/drive/MyDrive/Exam_2/facebook_event_checkin.csv
/content/drive/MyDrive/Exam_2/drivers_license.csv
/content/drive/MyDrive/Exam_2/crime_scene_report.csv


In [ ]:
person = pd.read_csv(person_file_path)
interview = pd.read_csv(interview_file_path)
income = pd.read_csv(income_file_path)
get_fit_now_member = pd.read_csv(get_fit_now_member_file_path)
get_fit_now_check_in = pd.read_csv(get_fit_now_check_in_file_path)
facebook_event_checkin = pd.read_csv(facebook_event_checkin_file_path)
driver_license = pd.read_csv(driver_license_file_path)
crime_scene_report = pd.read_csv(crime_scene_report_file_path)

In [ ]:
person

id                 name  license_id  address_number  \
0      10000   Christoper Peteuil      993845             624   
1      10007  Kourtney Calderwood      861794            2791   
2      10010            Muoi Cary      385336             741   
3      10016          Era Moselle      431897            1987   
4      10025         Trena Hornby      550890             276   
...      ...                  ...         ...             ...   
10006  99936          Luba Benser      274427             680   
10007  99941      Roxana Mckimley      975942            1613   
10008  99965      Cherie Zeimantz      287627            3661   
10009  99982          Allen Cruse      251350            3126   
10010  99990         Vance Hunten      830407            3056   

      address_street_name        ssn  
0            Bankhall Ave  747714076  
1           Gustavus Blvd  477972044  
2         Northwestern Dr  828638512  
3           Wood Glade St  614621061  
4           Daws Hill Way  223877684  
...                   ...        ...  
10006        Carnage Blvd  685095054  
10007             Gate St  512136801  
10008       The Water Ave  362877324  
10009           N Jean Dr  348734531  
10010       Lancefield St  896677562  

[10011 rows x 6 columns]

In [ ]:
interview

person_id                                         transcript
0         28508                ‘I deny it!’ said the March Hare.\n
1         63713                                                 \n
2         86208      way, and the whole party swam to the shore.\n
3         35267  lessons in here? Why, there’s hardly room for ...
4         33856                                                 \n
...         ...                                                ...
4986      37357  Alice did not wish to offend the Dormouse agai...
4987      10206  time,’ she said, ‘than waste it in asking ridd...
4988      14887  I heard a gunshot and then saw a man run out. ...
4989      16371  I saw the murder happen, and I recognized the ...
4990      67318  I was hired by a woman with a lot of money. I ...

[4991 rows x 2 columns]

In [ ]:
income

ssn  annual_income
0     100009868          52200
1     100169584          64500
2     100300433          74400
3     100355733          35900
4     100366269          73000
...         ...            ...
7509  999679296          54400
7510  999762859          77000
7511  999824984          82000
7512  999910617          82600
7513  999942603          11500

[7514 rows x 2 columns]

In [ ]:
get_fit_now_member

id  person_id                name  membership_start_date  \
0    NL318      65076     Everette Koepke               20170926   
1    AOE21      39426        Noe Locascio               20171005   
2    2PN28      63823  Jeromy Heitschmidt               20180215   
3    0YJ24      80651      Waneta Wellard               20171206   
4    3A08L      32858        Mei Bianchin               20170401   
..     ...        ...                 ...                    ...   
179  2V137      41693      Wendell Dulany               20171219   
180  4KB72      79110          Emile Hege               20170522   
181  48Z7A      28819        Joe Germuska               20160305   
182  48Z55      67318       Jeremy Bowers               20160101   
183  90081      16371      Annabel Miller               20160208   

    membership_status  
0                gold  
1             regular  
2              silver  
3                gold  
4              silver  
..                ...  
179            silver  
180           regular  
181              gold  
182              gold  
183              gold  

[184 rows x 5 columns]

In [ ]:
get_fit_now_check_in

membership_id  check_in_date  check_in_time  check_out_time
0            NL318       20180212            329             365
1            NL318       20170811            469             920
2            NL318       20180429            506             554
3            NL318       20180128            124             759
4            NL318       20171027            418            1019
...            ...            ...            ...             ...
2698         4KB72       20170422           1016            1114
2699         4KB72       20170630            408             885
2700         48Z7A       20180109           1600            1730
2701         48Z55       20180109           1530            1700
2702         90081       20180109           1600            1700

[2703 rows x 4 columns]

In [ ]:
facebook_event_checkin

person_id  event_id                                         event_name  \
0          28508      5880    Nudists are people who wear one-button suits.\n   
1          63713      3865  but that's because it's the best book on anyth...   
2          63713      3999           If Murphy's Law can go wrong, it will.\n   
3          63713      6436  Old programmers never die.  They just branch t...   
4          82998      4470               Help a swallow land at Capistrano.\n   
...          ...       ...                                                ...   
20006      99716      1143                               SQL Symphony Concert   
20007      99716      1143                               SQL Symphony Concert   
20008      99716      1143                               SQL Symphony Concert   
20009      67318      4719                             The Funky Grooves Tour   
20010      67318      1143                               SQL Symphony Concert   

           date  
0      20170913  
1      20171009  
2      20170502  
3      20170926  
4      20171022  
...         ...  
20006  20171206  
20007  20171212  
20008  20171229  
20009  20180115  
20010  20171206  

[20011 rows x 4 columns]

In [ ]:
driver_license

id  age  height eye_color hair_color  gender plate_number  \
0      100280   72      57     brown        red    male       P24L4U   
1      100460   63      72     brown      brown  female       XF02T6   
2      101029   62      74     green      green  female       VKY5KR   
3      101198   43      54     amber      brown  female       Y5NZ08   
4      101255   18      79      blue       grey  female       5162Z1   
...       ...  ...     ...       ...        ...     ...          ...   
10002  999923   19      77     amber      black  female       5L0ZI4   
10003  999940   71      61     green      green    male       1B8QN8   
10004  999981   67      69     brown       blue  female       1684K3   
10005  999986   49      58     green       grey    male       F8F64H   
10006  999993   18      63     black      black  female       6UZO2O   

         car_make    car_model  
0           Acura          MDX  
1        Cadillac          SRX  
2           Scion           xB  
3          Nissan        Rogue  
4           Lexus           GS  
...           ...          ...  
10002         GMC  Sierra 3500  
10003  Mitsubishi      Eclipse  
10004  Land Rover          LR2  
10005       Lexus           LS  
10006    Cadillac      DeVille  

[10007 rows x 9 columns]

In [ ]:
crime_scene_report

date     type                                        description  \
0     20180115  robbery  A Man Dressed as Spider-Man Is on a Robbery Spree   
1     20180115   murder                  Life? Dont talk to me about life.   
2     20180115   murder  Mama, I killed a man, put a gun against his he...   
3     20180215   murder                         REDACTED REDACTED REDACTED   
4     20180215   murder  Someone killed the guard! He took an arrow to ...   
...        ...      ...                                                ...   
1223  20180430  bribery                                                 \n   
1224  20180430    fraud                  ‘Why not?’ said the March Hare.\n   
1225  20180430  assault                                                 \n   
1226  20180501  assault  be NO mistake about it: it was neither more no...   
1227  20180115   murder  Security footage shows that there were 2 witne...   

              city  
0              NYC  
1           Albany  
2             Reno  
3         SQL City  
4         SQL City  
...            ...  
1223  Garden Grove  
1224         Houma  
1225       Fontana  
1226       Trenton  
1227      SQL City  

[1228 rows x 4 columns]



---



## เริ่มตรวจสอบข้อมูลที่ crime_scene_report

In [ ]:
# หาคดีฆาตกรรม (murder) ในวันที่ 15 มกราคม 2018 ที่เมือง SQL City
crime = crime_scene_report[(crime_scene_report.type == 'murder') & (crime_scene_report.date == 20180115) & (crime_scene_report.city == 'SQL City')]
crime


date    type                                        description  \
1227  20180115  murder  Security footage shows that there were 2 witne...   

          city  
1227  SQL City


> Security footage shows that there were 2 witnesses. The first witness lives at the last house on "Northwestern Dr". The second witness, named Annabel, lives somewhere on "Franklin Ave".







## ตรวจสอบข้อมูลพยาน

* find the first witness lives at the last house on "Northwestern Dr"

In [ ]:
# หาบ้านหลังสุดท้ายบนถนน Northwestern Dr
last_house_number = person[person.address_street_name == 'Northwestern Dr']['address_number'].max()

# หา person ที่อยู่บ้านหลังสุดท้ายบนถนน Northwestern Dr
witness1 = person[(person.address_street_name == 'Northwestern Dr') & (person.address_number == last_house_number)]
witness1


id            name  license_id  address_number address_street_name  \
499  14887  Morty Schapiro      118009            4919     Northwestern Dr   

           ssn  
499  111564949

In [ ]:
# หา interview ที่ person_id = 14887
interview_witness1 = interview[interview.person_id == 14887]
interview_witness1

person_id                                         transcript
4988      14887  I heard a gunshot and then saw a man run out. ...

> I heard a gunshot and then saw a man run out. He had a "Get Fit Now Gym" bag. The membership number on the bag started with "48Z". Only gold members have those bags. The man got into a car with a plate that included "H42W".




In [ ]:
# หาสมาชิก gold ที่มี id ขึ้นต้นด้วย '48Z'
get_fit_now_member[(get_fit_now_member.membership_status == 'gold') & (get_fit_now_member.id.str.startswith('48Z'))]

id  person_id           name  membership_start_date membership_status
181  48Z7A      28819   Joe Germuska               20160305              gold
182  48Z55      67318  Jeremy Bowers               20160101              gold

In [ ]:
person[person.id == 28819]

id          name  license_id  address_number address_street_name  \
2037  28819  Joe Germuska      173289             111             Fisk Rd   

            ssn  
2037  138909730

In [ ]:
person[person.id == 67318]

id           name  license_id  address_number    address_street_name  \
6327  67318  Jeremy Bowers      423327             530  Washington Pl, Apt 3A   

            ssn  
6327  871539279

In [ ]:
# find driver_license that a plate that included "H42W".
driver_license[driver_license.plate_number.str.contains('H42W')]

id  age  height eye_color hair_color  gender plate_number  \
915   183779   21      65      blue     blonde  female       H42W0X   
3529  423327   30      70     brown      brown    male       0H42W2   
6240  664760   21      71     black      black    male       4H42WR   

       car_make car_model  
915      Toyota     Prius  
3529  Chevrolet  Spark LS  
6240     Nissan    Altima

* find the second witness, named Annabel, lives somewhere on "Franklin Ave".

In [ ]:
# หา person ที่ชื่อ Annabel และอยู่  "Franklin Ave"
witness2 = person[(person.address_street_name == 'Franklin Ave') & (person.name.str.startswith('Annabel'))]
witness2


id            name  license_id  address_number address_street_name  \
665  16371  Annabel Miller      490173             103        Franklin Ave   

           ssn  
665  318771143

In [ ]:
# หา interview ที่ person_id = 16371
interview_witness2 = interview[interview.person_id == 16371]
interview_witness2

person_id                                         transcript
4989      16371  I saw the murder happen, and I recognized the ...



> I saw the murder happen, and I recognized the killer from my gym when I was working out last week on January the 9th.



In [ ]:
#หา checkindate == 20180209
get_fit_now_check_in[(get_fit_now_check_in.check_in_date == 20180109) & (get_fit_now_check_in.membership_id == '48Z55')]

membership_id  check_in_date  check_in_time  check_out_time
2701         48Z55       20180109           1530            1700

Jeremy Bowers
person id 67318
membership 48Z55
license_id 423327
age 30
hight 70
eyecolor brown
haircolor brown
gender male
plate number 0H42W2
car make Chevrolet
car model Spark LS


### คำให้การของ Jeremy Bowers

In [ ]:
interview_crime1 = interview[interview.person_id == 67318]
interview_crime1

person_id                                         transcript
4990      67318  I was hired by a woman with a lot of money. I ...



> I was hired by a woman with a lot of money. I don't know her name but I know she's around 5'5" (65") or 5'7" (67"). She has red hair and she drives a Tesla Model S. I know that she attended the SQL Symphony Concert 3 times in December 2017.



In [ ]:
# she's around 5'5" (65") or 5'7" (67"). She has red hair and she drives a Tesla Model S
driver_license[((driver_license.height.between(65,67))) & (driver_license.hair_color == 'red') & (driver_license.car_make == 'Tesla') & (driver_license.car_model == 'Model S')]


id  age  height eye_color hair_color  gender plate_number car_make  \
1105  202298   68      66     green        red  female       500123    Tesla   
2054  291182   65      66      blue        red  female       08CM64    Tesla   
9078  918773   48      65     black        red  female       917UU3    Tesla   

     car_model  
1105   Model S  
2054   Model S  
9078   Model S

In [ ]:
person[person.license_id == 202298]

id              name  license_id  address_number address_street_name  \
9985  99716  Miranda Priestly      202298            1883          Golden Ave   

            ssn  
9985  987756388

In [ ]:
income[income.ssn == 987756388]

ssn  annual_income
7422  987756388         310000

In [ ]:
income['annual_income'].max()

498500

In [ ]:
income['annual_income'].min()

10000

In [ ]:
person[person.license_id == 291182]

id           name  license_id  address_number address_street_name  \
8977  90700  Regina George      291182             332           Maple Ave   

            ssn  
8977  337169072

In [ ]:
income[income.ssn == 337169072]

Empty DataFrame
Columns: [ssn, annual_income]
Index: []

In [ ]:
person[person.license_id == 918773]

id      name  license_id  address_number address_street_name  \
7664  78881  Red Korb      918773             107         Camerata Dr   

            ssn  
7664  961388910

In [ ]:
income[income.ssn == 961388910]

ssn  annual_income
7207  961388910         278000

In [ ]:
# หาคนที่ check in SQL Symphony Concert  3 ครั้งในเดือน december ปี 2017
facebook_event_checkin[(facebook_event_checkin.event_name == 'SQL Symphony Concert') &
(facebook_event_checkin.date.between(20171201,20171231))].groupby('person_id').filter(lambda x: len(x)==3)

person_id  event_id            event_name      date
20001      24556      1143  SQL Symphony Concert  20171207
20002      24556      1143  SQL Symphony Concert  20171221
20003      24556      1143  SQL Symphony Concert  20171224
20006      99716      1143  SQL Symphony Concert  20171206
20007      99716      1143  SQL Symphony Concert  20171212
20008      99716      1143  SQL Symphony Concert  20171229

In [ ]:
person[person.id == 24556]

id         name  license_id  address_number address_street_name  \
1587  24556  Bryan Pardo      101191             703          Machine Ln   

            ssn  
1587  816663882

In [ ]:
driver_license[driver_license.id == 101191]

Empty DataFrame
Columns: [id, age, height, eye_color, hair_color, gender, plate_number, car_make, car_model]
Index: []

In [ ]:
income[income.ssn == 816663882]

Empty DataFrame
Columns: [ssn, annual_income]
Index: []

In [ ]:
person[person.id == 99716]

id              name  license_id  address_number address_street_name  \
9985  99716  Miranda Priestly      202298            1883          Golden Ave   

            ssn  
9985  987756388

In [ ]:
driver_license[driver_license.id == 202298]

id  age  height eye_color hair_color  gender plate_number car_make  \
1105  202298   68      66     green        red  female       500123    Tesla   

     car_model  
1105   Model S

In [ ]:
income[income.ssn == 987756388]

ssn  annual_income
7422  987756388         310000

In [ ]:
interview_crime2 = interview[interview.person_id == 99716]
interview_crime2

Empty DataFrame
Columns: [person_id, transcript]
Index: []



---



Miranda Priestly (ผู้ว่าจ้าง)

id 99716
license_id 202298
address_number 1883
address_street_name	Golden Ave
ssn 987756388
age	68
height 66
eye_color	green
hair_color	red
gender female
plate_number	500123
car_make	Tesla
car_model Model S
income 310000


Jeremy Bowers (ถูกว่าจ้าง)

membership 48Z55
license_id 423327
age 30                     
hight 70 eyecolor brown
haircolor brown gender male plate number 0H42W2
car make Chevrolet
car model



---

